### PR TITLE
Load and build multiple universes in ORANGE

### PR DESCRIPTION
**Pull request by elliottbiondo**

### CODE REVIEW

**Comment by elliottbiondo** [comment]
@sethrj this is ready for your review

**Comment by sethrj** [comment]
@elliottbiondo seeing a bunch of test failures? https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Celeritas/detail/PR-534/11/tests

**Comment by elliottbiondo** [comment]
Yes I am now --- I must have forgot to run the full suite after rebasing. These weren't failing as of last night. Will investigate.

**Comment by elliottbiondo** [comment]
Okay these might have broken earlier without realizing (thinking they were unrelated). It looks like many of them come from the changes I made from volume naming, which affect material assignments as well. Before I go through the full code base, re you okay with the `universe.volume.surface` naming scheme?

**Comment by sethrj** [comment]
@elliottbiondo The namespacing should probably be a "fallback" if there are duplicates. Usually when people want to reference a unique object they don't give the full hierarchy to get to that object: in Shift we try to use the labels that users provide rather than scoping them.

To make things more complicated, we need to make sure the volume names can be represented the same way between ORANGE and VecGeom...

**Comment by elliottbiondo** [comment]
@sethrj let's talk about this tomorrow; sounds like there are quite a few requirements on the labelling scheme.

File: src/orange/detail/UnitIndexer.hh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    //! Total number of surfaces
+    size_type num_surfaces() const { return d_surfaces.back(); }
+
+    //! Total number of cells
+    size_type num_volumes() const { return d_volumes.back(); }
+
+  private:
+    // >>> DATA
+    Vec_Size d_surfaces;
+    Vec_Size d_volumes;
```
Comment (sethrj) [inline src/orange/detail/UnitIndexer.hh:58]: Some stylistic changes versus what we do in Shift (avoid the diff markers `>>>` and `<<<`, use trailing underscore for member data, no uppercase+underscore in type names)

```suggestion
    //// DATA ////
    VecSize surfaces_;
    VecSize volumes_;
```

File: src/orange/detail/UnitInserter.cc [hunk 2]
```diff
@@ -338,6 +351,18 @@ VolumeRecord UnitInserter::insert_volume(const SurfacesRecord& surf_record,
     return output;
 }
 
+void UnitInserter::process_daughter(VolumeRecord&              vol_record,
+                                    std::vector<Translation>   translations,
```
Comment (sethrj) [inline src/orange/detail/UnitInserter.cc:355]: This parameter is mistakenly by value (oops!)

Also in Celeritas to make it obvious from a calling site that an argument is expected to be modified, we pass raw pointers rather than mutable references: this was the recommendation of the google style guide up to early 2020 (I just now found out it's changed). The practice would have made it obvious here that a reference is missing on the `translations` argument.

File: test/orange/Orange.test.cc [hunk 3]
```diff
@@ -85,8 +91,8 @@ TEST_F(OneVolumeTest, params)
     EXPECT_EQ(0, geo.num_surfaces());
     EXPECT_TRUE(geo.supports_safety());
 
-    EXPECT_EQ("infinite", geo.id_to_label(VolumeId{0}).name);
-    EXPECT_EQ(VolumeId{0}, geo.find_volume("infinite"));
+    EXPECT_EQ("one_volume.infinite", geo.id_to_label(VolumeId{0}).name);
+    EXPECT_EQ(VolumeId{0}, geo.find_volume("one_volume.infinite"));
```
Comment (sethrj) [inline test/orange/Orange.test.cc:95]: As per our pending discussion we'll want to revert these changes.

File: test/orange/Orange.test.cc [hunk 4]
```diff
@@ -502,6 +508,27 @@ TEST_F(FiveVolumesTest, params)
...
+
+    EXPECT_EQ("outer.[EXTERIOR]", geo.id_to_label(VolumeId{0}).name);
+    EXPECT_EQ("outer.inner_a", geo.id_to_label(VolumeId{1}).name);
+    EXPECT_EQ("outer.inner_b", geo.id_to_label(VolumeId{2}).name);
+    EXPECT_EQ("outer.bobby", geo.id_to_label(VolumeId{3}).name);
+    EXPECT_EQ("outer.johnny", geo.id_to_label(VolumeId{4}).name);
+    EXPECT_EQ("inner.[EXTERIOR]", geo.id_to_label(VolumeId{5}).name);
+    EXPECT_EQ("inner.a", geo.id_to_label(VolumeId{6}).name);
+    EXPECT_EQ("inner.b", geo.id_to_label(VolumeId{7}).name);
+    EXPECT_EQ("inner.c", geo.id_to_label(VolumeId{8}).name);
```
Comment (sethrj) [inline test/orange/Orange.test.cc:529]: This is all fine, but it might've been easier (and will be more robust in case the number of volumes is wrong) to loop over `range(VolumeId{geo.num_volumes()})` , `push_back` onto a vector of strings, and use `EXPECT_VEC_EQ`. (You can use the `PRINT_EXPECTED` macro to generate the initial list of expected values.)

File: test/orange/OrangeGeoTestBase.cc [hunk 5]
```diff
@@ -117,7 +117,7 @@ void OrangeGeoTestBase::build_geometry(OneVolInput inp)
     // Save fake bbox for sampling
     input.bbox = {{-0.5, -0.5, -0.5}, {0.5, 0.5, 0.5}};
 
-    input.label = "one volume";
+    input.label = "one_volume";
```
Comment (sethrj) [inline test/orange/OrangeGeoTestBase.cc:120]: Spaces should be allowable as part of a label, so please revert this.

File: test/orange/data/universes.org.json [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+"bob.px",
+"inner_a.my",
+"bob.my",
+"bob.mz",
+"inner_a.pz",
+"bob.pz",
+"bob.py"
+],
+"surfaces": {
+"connectivity": [
```
Comment (sethrj) [inline test/orange/data/universes.org.json:144]: Since we now generate connectivity at construction time, can you delete this from the JSON file? And in Shift ORANGE on your branch, can you likewise delete the connectivity from the output routines?

**Review by sethrj — CHANGES_REQUESTED** [review]
First round of feedback: overall the new code looks good but as you said we need to talk about the global labeling.

**Comment by sethrj** [comment]
Also @elliottbiondo a few small pointers about the PR submission itself (since this might be one of the first GitHub submissions you've done in a long time):
- When you first submit a PR, there's a drop-down option on the "submit" box that marks it as a "draft" straight away. You can omit `WIP` from the PR title when it's a draft.
- The title should be formatted like a git commit title: a present tense action describing the change. Try to be both descriptive and concise, since these PR titles go into the release notes.
- When you're ready for review there's another little button on the right hand side (hard to find) that removes the draft mark.
- For PRs that touch (rougly) more than a single file, the description should summarize the changes you've made. In this case you should add  something about input processing and translations along with a caveat that multi-universe transport is still not enabled.

File: src/orange/OrangeParams.cc [hunk 7]
```diff
@@ -125,22 +138,31 @@ OrangeParams::OrangeParams(OrangeInput input)
...
+        for (const auto& sl : u.surfaces.labels)
+        {
+            surface_labels.push_back(u.label.name + std::string(".") + sl.name);
+        }
+
+        for (const auto& v : u.volumes)
         {
-            volume_labels[i] = std::move(u.volumes[i].label);
+            volume_labels.push_back(u.label.name + std::string(".")
+                                    + v.label.name);
```
Comment (sethrj) [inline src/orange/OrangeParams.cc:156]:
```suggestion
            Label vol_label{v.label.name, u.label.name};
            if (v.label.ext)
            {
                vol_label.ext += "." + v.label.ext;
            }
            volume_labels.push_back(vol_label);
```

Comment (elliottbiondo) [reply src/orange/OrangeParams.cc:156]: Same thing for surface labels I pressume?

Comment (elliottbiondo) [reply src/orange/OrangeParams.cc:156]: Wait, if aren't using the `UniverseId` to uniqify and we are preserving the `ext` from the input (if it exists) shouldn't we just be doing:
```
        for (const auto& v : u.volumes)
        {
            volume_labels.push_back(v.label);
        }
```

Comment (elliottbiondo) [reply src/orange/OrangeParams.cc:156]: (as a copy with an lambda function)

**Comment by elliottbiondo** [comment]
@sethrj this is ready for another look.

**Comment by sethrj** [comment]
@elliottbiondo Please don't force push after a review has started. Can you take a quick look at the [contributing guidelines](https://github.com/celeritas-project/celeritas/blob/master/CONTRIBUTING.rst) just to take a look at what differs between Shift and Celeritas PRs? Thanks.

Screen Shot 2022-10-28 at 16 34 10

File: src/orange/OrangeParams.cc (hunk 7 above)
Comment (sethrj) [reply src/orange/OrangeParams.cc]: I thought we decide to use the universe name to uniquify via the extension if the incoming label lacks an extension? So if the input is `"a@b"` it turns into `{"a", "b"}`, but if it's `"a"` in universe 'u' then it becomes `{"a", "u"}`

File: src/orange/OrangeParams.cc [hunk 8]
```diff
@@ -97,11 +114,10 @@ OrangeParams::OrangeParams(const std::string& json_filename)
...
  */
 OrangeParams::OrangeParams(OrangeInput input)
+    : unit_indexer_(make_unit_indexer(input.units))
 {
-    CELER_VALIDATE(input.units.size() == 1,
-                   << "input geometry has " << input.units.size()
-                   << "universes; at present there must be a single global "
-                      "universe");
+    CELER_VALIDATE(!input.units.empty(),
+                   << "input geometry has must contain at least one universe");
```
Comment (sethrj) [inline src/orange/OrangeParams.cc:120]: Minor comment, but I've tried to be very consistent about how we write error messages (since they're always presented as `error: <blah>`): see https://github.com/celeritas-project/celeritas/blob/95fb1e9c8e72acba7800f053ccbd4887a64622e0/src/corecel/Assert.hh#L52 .

```suggestion
                   << "input geometry has no universes");
```
(only after editing this did I notice `has must` 😄 )

File: src/orange/OrangeParams.cc [hunk 9]
```diff
@@ -125,22 +141,30 @@ OrangeParams::OrangeParams(OrangeInput input)
...
-        for (auto i : range(u.volumes.size()))
-        {
-            volume_labels[i] = std::move(u.volumes[i].label);
-        }
-        vol_labels_ = LabelIdMultiMap<VolumeId>{std::move(volume_labels)};
-        bbox_       = u.bbox;
+
+        std::copy(u.surfaces.labels.begin(),
+                  u.surfaces.labels.end(),
+                  std::back_inserter(surface_labels));
```
Comment (sethrj) [inline src/orange/OrangeParams.cc:153]: Nice use of algorithm! Better yet, though:
```suggestion
        surface_labels.insert(surface_labels.insert .end(),
                u.surfaces.labels.begin(),
                u.surfaces.labels.end());
```

File: src/orange/OrangeParams.cc [hunk 10]
```diff
@@ -125,22 +141,30 @@ OrangeParams::OrangeParams(OrangeInput input)
...
+
+        std::transform(u.volumes.begin(),
+                       u.volumes.end(),
+                       std::back_inserter(volume_labels),
+                       [](VolumeInput v) -> Label { return v.label; });
```
Comment (sethrj) [inline src/orange/OrangeParams.cc:158]: See the comment about volume labels: I thought we decided to uniquify on universe name if no "ext" is provided?

File: src/orange/OrangeParams.cc (hunk 7 above)
Comment (elliottbiondo) [reply src/orange/OrangeParams.cc]: I think I got turned around on this --- I will change it to the above code. I will do the same for surfaces unless I hear otherwise.

**Comment by elliottbiondo** [comment]
Okay @sethrj this should be ready

**Comment by sethrj** [comment]
@elliottbiondo CI has caught some typos:
```
/var/jenkins/workspace/Celeritas_PR-534/src/orange/OrangeParams.cc:154:30: error: 'unit' was not declared in this scope; did you mean 'uint'?
  154 |         for (const auto& s : unit.sufurfaces.labels)
      |                              ^~~~
      |                              uint
/var/jenkins/workspace/Celeritas_PR-534/src/orange/OrangeParams.cc:164:30: error: 'unit' was not declared in this scope; did you mean 'uint'?
  164 |         for (const auto& v : unit.volumes)
      |                              ^~~~
      |                              uint
```

**Comment by elliottbiondo** [comment]
Oh gosh, sorry about that. I fixed the obvious issues, but it looks like there are two additional failures, probably stemming from the volume_id -> materials mapping. I'll get to those Monday.

**Comment by sethrj** [comment]
No problem!

**Comment by elliottbiondo** [comment]
@sethrj take a look at my final 2-line commit. I am not familiar enough with the full code base yet to know if this is going to be suitable for VecGeom compatibility. However, I can say that all non-disabled tests are passing locally.

File: src/orange/detail/UnitIndexer.cc [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+    UniverseId uni(iter - surfaces_.begin());
+    SurfaceId  surface(id - *iter);
+    CELER_ENSURE(uni.get() < this->num_universes());
+    return std::make_tuple(uni, surface);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Global-to-local
```
Comment (sethrj) [inline src/orange/detail/UnitIndexer.cc:104]:
```suggestion
 * Transform global to local volume ID.
```

File: src/orange/detail/UnitIndexer.cc [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    VolumeId   volume(id - *iter);
+    CELER_ENSURE(uni.get() < this->num_universes());
+    return std::make_tuple(uni, volume);
+}
+
+//---------------------------------------------------------------------------//
+// IMPLEMENTATION METHODS
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Locate the given ID in the list of offsets.
```
Comment (sethrj) [inline src/orange/detail/UnitIndexer.cc:121]:
```suggestion
 * Locate the given ID in the list of offsets.
```

File: src/orange/detail/UnitIndexer.cc [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    --iter;
+
+    CELER_ENSURE(*iter <= id);
+    CELER_ENSURE(id < *(iter + 1));
+    return iter;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Get the number of elements in the given universe.
```
Comment (sethrj) [inline src/orange/detail/UnitIndexer.cc:140]:
```suggestion
 * Get the number of elements in the given universe.
```

File: src/orange/detail/UnitIndexer.hh [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+    //! Total number of cells
+    size_type num_volumes() const { return volumes_.back(); }
+
+  private:
+    //// DATA ////
+    VecSize surfaces_;
+    VecSize volumes_;
+
+    // >>> IMPLEMENTATION METHODS
```
Comment (sethrj) [inline src/orange/detail/UnitIndexer.hh:60]:
```suggestion
    //// IMPLEMENTATION METHODS ////
```

File: test/celeritas/field/FieldPropagator.test.cc [hunk 15]
```diff
@@ -693,7 +693,7 @@ TEST_F(TwoBoxTest, electron_tangent_cross)
         if (!CELERITAS_USE_VECGEOM)
         {
             EXPECT_EQ("inner_box.py",
-                      this->geometry()->id_to_label(geo.surface_id()));
+                      this->geometry()->id_to_label(geo.surface_id()).name);
```
Comment (sethrj) [inline test/celeritas/field/FieldPropagator.test.cc:696]: It might be worth adding a `surface_name(geo)` test fixture accessor like the `volume_name` on line 699, for brevity.

File: test/celeritas/field/FieldPropagator.test.cc [hunk 16]
```diff
@@ -837,8 +837,8 @@ TEST_F(TwoBoxTest, electron_step_endpoint)
         Real3     start_pos{-5 + dx, 0, 0};
         axpy(real_type(-1), first_pos, &start_pos);
 
-        auto geo       = this->init_geo(start_pos, {0, 1, 0});
-        auto stepper   = make_mag_field_stepper<DiagnosticDPStepper>(
+        auto geo     = this->init_geo(start_pos, {0, 1, 0});
+        auto stepper = make_mag_field_stepper<DiagnosticDPStepper>(
```
Comment (sethrj) [inline test/celeritas/field/FieldPropagator.test.cc:841]: I think we must be using different clang-format versions... :( I want to get rid of the equal sign indentation for this reason.

**Review by sethrj — CHANGES_REQUESTED** [review]

**Comment by sethrj** [comment]
@elliottbiondo FYI with github, you can go to the "files" view and click "add suggestion to batch" for all the code change recommendations, then commit them in one go.

File: test/celeritas/field/FieldPropagator.test.cc (hunk 15 above)
Comment (elliottbiondo) [reply test/celeritas/field/FieldPropagator.test.cc:696]: Looks it already exists, but wasn't being used for some reason

File: test/celeritas/field/FieldPropagator.test.cc (hunk 16 above)
Comment (elliottbiondo) [reply test/celeritas/field/FieldPropagator.test.cc:841]: I am not sure I understand, do you prefer how it is in green?

**Comment by elliottbiondo** [comment]
@sethrj this is ready for another look

File: test/celeritas/field/FieldPropagator.test.cc (hunk 16 above)
Comment (sethrj) [reply test/celeritas/field/FieldPropagator.test.cc:837]: I personally care less how it's formatted and more that the line doesn't change if the code doesn't change. 😞  which is the case here... don't worry about reverting it though. The point of an autoformatter is to save us from having to deal with spaces 😬

**Review by sethrj — CHANGES_REQUESTED** [review]
One last little round... thanks for your patience with my comments; the earliest reviews are always the most difficult!

**Comment by elliottbiondo** [comment]
@sethrj I think this is good-to-go

**Review by sethrj — APPROVED** [review]
Nice, thanks for iterating with me Elliott!

File: src/orange/construct/OrangeInput.hh [hunk 17]
```diff
@@ -78,9 +80,17 @@ struct VolumeInput
  */
 struct UnitInput
 {
+    struct Daughter
+    {
+        UniverseId  universe_id;
+        Translation translation;
+    };
+    using Daughter_Map = std::unordered_map<VolumeId, Daughter>;
```
Comment (sethrj) [inline src/orange/construct/OrangeInput.hh:88]: I missed the naming issue here: there should be no underscores in type names, and for maps we're so far sticking to `MapKeyValue`. I'll rename this to MapVolumeDaughter.
